src/Filters/Filter.h

```diff
@@ -0,0 +1,18 @@
+#ifndef FILTER_H
+#define FILTER_H
+
+class Filter {
+public:
+    virtual ~Filter() {}
+
+    // Define core interface
+    virtual void initialize() = 0;
+    virtual double* iterate(double dt, double* state, double* measurements, double* controlVars) = 0;
+    
+    // Provide interface to query filter dimensions
+    virtual int getMeasurementSize() const = 0;
+    virtual int getInputSize() const = 0;
```

Comment:
why not one for the control vars size?

Reply:
That is what `getInputSize()` does, I guess it's just poorly named

Reply:
@mallamacimj With this, is this an appropriate function to have at the `Filter` class level? Even if renamed to like `getControlSize()`, this is used at the Filter level by the `State` class to get the size to "allocate" for the **U** vector, but isn't this a very KF-specific functionality? Say we had a different filtering algorithm that didn't necessarily have controls like that, then this method would not apply even if it fits under the "Filter" umbrella.

With that, should it be that `State` directly uses/includes `LinearKalmanFilter` class instead of the `Filter` class, if we do the above change? The way `State` is currently formatted and written, it works well with only KFs, and expects there to be a control vector, and so if we change `Filter` to not have that, then it would need to use the KF at the `LinearKalmanFilter` level. This would be unless we wanted another layer of abstraction above this as just `KalmanFilter` in general, and this is where a function like `getControlSize()` would live. Or if that's too much and we keep it, I guess that means we just treat `Filter` as effectively being a `KalmanFilter` class by it having this method signature, but not explicitly prohibiting it from being some other filtering algorithm (just by the sake of its naming, if nothing else).

Reply:
out state class is wrong. it is too hardcoded to a specific filter. how I did this in the past and it decouples the code better is having the end user have to define and use the filter as its own object. state was just a way to store the data. For example the payload main file had this in the loop

```
    //Run the KF
    dt = TADPOLSTATE.timeAbsolute - KF_PREVIOUS_ITER_TIME;
    double* meas_arr = new double[3] {TADPOLSTATE.stateGPS.gpsPosition.x(), TADPOLSTATE.stateGPS.gpsPosition.y(), TADPOLSTATE.stateBarometer.relativeAltitude};
    double* control_arr = new double[3] {TADPOLSTATE.stateIMU.acceleration.x(), TADPOLSTATE.stateIMU.acceleration.y(), TADPOLSTATE.stateIMU.acceleration.z() - 9.8};
    LinearKFConfig = iterate_filter(dt, meas_arr, control_arr, TADPOLSTATE.stateGPS.satellites);
    Matrix X = kf.iterate(LinearKFConfig);

    double* KFOutputState = X.getArr();
    // //Set the kf output data in the TADPOLSTATE 
    TADPOLSTATE.position.x() = KFOutputState[0]; TADPOLSTATE.position.y() = KFOutputState[1]; TADPOLSTATE.position.z() = KFOutputState[2];
    TADPOLSTATE.velocity.x() = KFOutputState[3]; TADPOLSTATE.velocity.y() = KFOutputState[4]; TADPOLSTATE.velocity.z() = KFOutputState[5];
    TADPOLSTATE.stateIMU.acceleration.x() = TADPOLSTATE.stateIMU.acceleration.x(); TADPOLSTATE.acceleration.y() = TADPOLSTATE.stateIMU.acceleration.y(); TADPOLSTATE.acceleration.z() = TADPOLSTATE.stateIMU.acceleration.z();
```

https://github.com/Terrapin-Rocket-Team/SAC-TRT24/blob/main/Code/Payload/TADPOL_SAC_2024/TADPOL_SAC_2024.ino

What we do can be cleaner than this ^^. However, we don't have to attach the filter to the state. Our main loops should read something like this

1. Read sensor data
2. Filter sensor data
3. Run data through controller to get actuator outputs
4. Send actuators commands
5. Record Data (sensor outputs, filter outputs, data sent to controller)

Reply:
Imo I think filter should be a part of state in the way we discussed during that one meeting we had towards the end of Summer. filter should be able to represent any filter (kf or whatever else we use) and can have a method iterate (or predict, just anything that gets estimated state values) and method(s) to actually input data. My main reasoning behind this is that the filters help us get a more accurate version of our state, so it should be a part of the state class. I'm not opposed to decoupling it though, it just seems slightly less clean.